controller/sync.go

```diff
@@ -112,253 +115,326 @@ func (m *appStateManager) SyncAppState(app *appv1.Application, state *appv1.Oper
+					case HealthStatusHealthy:
+						sc.setResourceResult(task, task.syncStatus, OperationSucceeded, healthStatus.Message)
+					case HealthStatusDegraded:
+						sc.setResourceResult(task, task.syncStatus, OperationFailed, healthStatus.Message)
+					}
+				}
 			}
-			// If apply was successful, return here and force an app refresh. This is so the app
-			// will become requeued into the workqueue, to force a new sync/health assessment before
-			// marking the operation as completed
-			return
-		}
-		sc.setOperationPhase(appv1.OperationSucceeded, "successfully synced")
-	} else if sc.syncOp.SyncStrategy.Hook != nil {
-		hooks, err := sc.getHooks()
-		if err != nil {
-			sc.setOperationPhase(appv1.OperationError, fmt.Sprintf("failed to generate hooks resources: %v", err))
-			return
 		}
-		sc.doHookSync(syncTasks, hooks)
-	} else {
-		sc.setOperationPhase(appv1.OperationFailed, "Unknown sync strategy")
+	}
+
+	// any running tasks, lets wait...
+	if tasks.Find(func(t *syncTask) bool { return t.running() }) != nil {
+		sc.setOperationPhase(OperationRunning, "one or more tasks are running")
+		return
+	}
+
+	// if there are any completed but unsuccessful tasks, sync is a failure.
+	if tasks.Find(func(t *syncTask) bool { return t.completed() && !t.successful() }) != nil {
+		sc.setOperationPhase(OperationFailed, "one or more synchronization tasks completed unsuccessfully")
 		return
 	}
+
+	sc.log.WithFields(log.Fields{"tasks": tasks}).Debug("filtering out completed tasks")
+	// remove tasks that are completed, we can assume that there are no running tasks
+	tasks = tasks.Filter(func(t *syncTask) bool { return !t.completed() })
+
+	// If no sync tasks were generated (e.g., in case all application manifests have been removed),
+	// the sync operation is successful.
+	if len(tasks) == 0 {
+		sc.setOperationPhase(OperationSucceeded, "successfully synced")
+		return
+	}
+
+	// remove any tasks not in this wave
+	phase := tasks.phase()
+	wave := tasks.wave()
+
+	sc.log.WithFields(log.Fields{"phase": phase, "wave": wave, "tasks": tasks}).Debug("filtering tasks in correct phase and wave")
+	tasks = tasks.Filter(func(t *syncTask) bool { return t.phase == phase && t.wave() == wave })
+
+	sc.setOperationPhase(OperationRunning, fmt.Sprintf("running phase='%s' wave=%d", phase, wave))
+
+	sc.log.WithFields(log.Fields{"tasks": tasks}).Debug("wet-run")
+	if !sc.runTasks(tasks, false) {
+		sc.setOperationPhase(OperationFailed, "one or more objects failed to apply")
+	}
 }
 
-// generateSyncTasks() generates the list of sync tasks we will be performing during this sync.
-func (sc *syncContext) generateSyncTasks() ([]syncTask, bool) {
-	syncTasks := make([]syncTask, 0)
-	successful := true
-	for _, resourceState := range sc.compareResult.managedResources {
-		if resourceState.Hook {
+func (sc *syncContext) notStarted() bool {
+	return len(sc.syncRes.Resources) == 0
+}
+
+func (sc *syncContext) isSelectiveSync() bool {
+	return sc.syncResources != nil
```

Comment:
This is odd, if you sync all resources from the UI, then `len(sc.syncResources) > 0`, but this is not selective.